nimble/controller/src/ble_ll_conn.c

```diff
@@ -3295,21 +3310,57 @@ ble_ll_init_rx_isr_end(uint8_t *rxbuf, uint8_t crcok,
                 !inita_is_rpa) {
                 goto init_rx_isr_exit;
             }
+
+            /*
+             * If the InitA is a RPA, we must see if it resolves based on the
+             * identity address of the resolved ADVA.
+             */
+            if (init_addr && inita_is_rpa &&
+                            !ble_ll_resolv_rpa(init_addr,
+                                               g_ble_ll_resolv_list[index].rl_local_irk)) {
+                goto init_rx_isr_exit;
+            }
+
         } else {
             if (chk_wl) {
                 goto init_rx_isr_exit;
             }
+
+            /* Could not resolved InitA */
+            if (init_addr && inita_is_rpa) {
+                goto init_rx_isr_exit;
+            }
         }
-    } else if (init_addr && ble_ll_resolv_enabled()) {
+    } else if (init_addr) {
 
-        /* Let's see if we have IRK with that peer. If so lets make sure
-         * privacy mode is correct together with initA
-         */
+        /* If resolving is off and InitA is RPA we reject advertising */
+        if (!ble_ll_resolv_enabled() && inita_is_rpa) {
```

Comment:
it's better to check variable first since if it evaluates to false there is no need to call function